server/legalhold/legal_hold.go

```diff
@@ -40,42 +46,84 @@ type Execution struct {
 }
 
 // NewExecution creates a new Execution that is ready to use.
-func NewExecution(legalHold model.LegalHold, papi plugin.API, store *sqlstore.SQLStore, fileBackend filestore.FileBackend) Execution {
+func NewExecution(legalHold model.LegalHold, papi plugin.API, store *sqlstore.SQLStore, kvstore kvstore.KVStore, fileBackend filestore.FileBackend) Execution {
 	return Execution{
 		LegalHold:          legalHold,
 		ExecutionStartTime: legalHold.NextExecutionStartTime(),
 		ExecutionEndTime:   legalHold.NextExecutionEndTime(),
 		store:              store,
+		kvstore:            kvstore,
 		fileBackend:        fileBackend,
 		index:              model.NewLegalHoldIndex(),
 		papi:               papi,
 		hashes:             make(map[string]string),
 	}
 }
 
-// Execute executes the Execution.
-func (ex *Execution) Execute() (int64, error) {
-	err := ex.GetChannels()
+// Execute executes the Execution and returns the updated LegalHold.
+func (ex *Execution) Execute() (*model.LegalHold, error) {
+	now := time.Now().Unix()
+
+	// Lock multiple executions using a cluster mutex
+	mutexKey := fmt.Sprintf("legal_hold_%s_execution", ex.LegalHold.ID)
+	mutex, err := cluster.NewMutex(ex.papi, mutexKey)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create cluster mutex: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), executionGlobalTimeout)
```

Comment:
48 hour timeout is basically no timeout.  I think it would be better to timeout sooner, provided the timeout simply stops collecting data and picks up where it left off next run.

Reply:
Right now the context is only being used for the cluster mutex since no other operation from the legal hold executor uses it. Modifying it properly to handle a context cancellation would require significant changes and storing the state would depend on which step of the job stops. I can open a ticket to track this.

Reply:
https://github.com/mattermost/mattermost-plugin-legal-hold/issues/133

Reply:
I think the idea here is to fail immediately if another execution is already running.  Waiting 2 hours means someone triggering a manual run via UI or slash command will get no feedback for 2 hours.  A quick failure, and error message back to the user will let them try again later.  Ideally we'll eventually have a status dashboard in the system console config area where they can see a job is running and what percentage complete it is.

IMO a 5 or 10 second timeout is what is needed here.  Calling code can decide if/when to try again.

Thoughts?

Reply:
Yeah you are right, I kept thinking the context was for the running job while is only for the mutex, even when I was saying so. Let's rename the variables and lower the timeout to 5 seconds.